--- thread-index-mode.rb~	2011-04-21 16:36:25.000000000 +1000
+++ thread-index-mode.rb	2011-06-16 13:08:23.000000000 +1000
@@ -43,6 +43,7 @@
     k.add :toggle_spam, "Mark/unmark thread as spam", 'S'
     k.add :toggle_deleted, "Delete/undelete thread", 'd'
     k.add :kill, "Kill thread (never to be seen in inbox again)", '&'
+    k.add :save_thread_to_disk, "Save thread to disk", 's'
     k.add :flush_index, "Flush all changes now", '$'
     k.add :jump_to_next_new, "Jump to next new thread", :tab
     k.add :reply, "Reply to latest message in a thread", 'r'
@@ -465,6 +466,28 @@
     multi_kill [t]
   end
 
+  def save_thread_to_disk
+    # Apologies if this is crap, I don't know Ruby at all
+    t = cursor_thread or return
+    dir = BufferManager.ask_for_filename :filename, "Save messages to directory: "
+    return unless dir
+    if File.exists? dir
+      return
+    end
+    Dir.mkdir dir
+    t.each do |m, depth, parent|
+      next unless m
+      fn = File.expand_path File.join(dir, m.subj.gsub(/\//, "_"))
+      while File.exists? (fn+".mbox")
+        fn = fn+"#"
+      end
+      fn = fn+".mbox"
+      save_to_file(fn) do |f|
+        m.each_raw_message_line { |l| f.print l }
+      end
+    end
+  end
+
   def flush_index
     @flush_id = BufferManager.say "Flushing index..."
     Index.save_index
